simulator/func_sim/trap_types.h

```diff
@@ -8,18 +8,54 @@
 
 // MIPT-MIPS modules
 #include <infra/types.h>
+#include "riscv.opcode.gen.h"
 
-enum class Trap : uint8
-{
-    NO_TRAP,
-    HALT,
-    EXPLICIT_TRAP,
-    BREAKPOINT,
-    SYSCALL,
-    INTEGER_OVERFLOW,
-    UNSUPPORTED_SYSCALL,
-    UNALIGNED_ADDRESS,
-    UNKNOWN_INSTRUCTION
+#include <array>
+#include <string_view>
+
+
+class Trap {
+    public:
+        enum TrapType : uint8
+        {
+            //NOLINTNEXTLINE(cppcoreguidelines-macro-usage)
+            #define TRAP(name) name,
+            #include <func_sim/trap_types.def>
+            #undef TRAP
+            MAX_TRAP_TYPE
+        };
+
+        constexpr Trap(TrapType id) : value(id) { }
+
+        operator uint8() const { return value; }
```

Comment:
Not sure whether we need that method. For hash, you may create a method `get_hash` returning a hashed value.

Reply:
Earlier there were two comparison operators. I created this method instead of get value/hash and operator!=, operator ==.

Reply:
Explicit constructor will not work without comparsion operators. I'll remove uint8 type cast.

Reply:
Explicit comparison operators are better in that case.
With implicit conversion, you're losing control what may happen with that type.

Reply:
I don't know how to use explicit so as not to change gdb-wrapper code. Can I edit it and replace old constructions with new methods?

Reply:
Yes, sure.

Reply:
I think that the methods proposed in the last commit would be the best option.
And only conversion operators and constructors may have explicit qualifier.